tools/autograd/load_derivatives.py

```diff
@@ -87,11 +89,175 @@ def create_derivative(f: NativeFunction, formula: str, var_names: Tuple[str, ...
+
+def postprocess_forward_derivatives(
+    f: NativeFunction,
+    defn_name: str,
+    all_arg_names: List[str],
+    derivatives: List[Derivative],
+    forward_derivatives: List[ForwardDerivative],
+    args_with_derivatives: Sequence[Binding]
+) -> List[ForwardDerivative]:
+
+    def find_required_inputs(formula: str, postfix: str) -> Tuple[str, ...]:
+        required_inputs = set()
+        for arg in args_with_derivatives:
+            if arg.type == 'TensorList':
+                # The functions taking TensorList handle everything internally
+                continue
+            arg_name = arg.name
+
+            found = re.search(IDENT_REGEX.format(arg_name), formula)
+            if found:
+                raise RuntimeError(f"The forward formula for {defn_name} is using the base name of the {arg_name} "
+                                   f"argument which is ambiguous. You should use {arg_name}_p to access the primal "
+                                   f"value and {arg_name}_t to access the tangent.")
+
+            found = re.search(IDENT_REGEX.format(arg_name + postfix), formula)
+            if found:
+                required_inputs.add(arg_name)
+
+        return tuple(required_inputs)
+
+    updated_derivatives: List[ForwardDerivative] = []
+
+    for defn in forward_derivatives:
+        formula = defn.formula
+        required_inputs_tangent = find_required_inputs(formula, "_t")
+        if formula == "auto_element_wise":
+            if (not len(args_with_derivatives) == 1) or len(forward_derivatives) > 1:
+                raise RuntimeError(f"Derivative definition of {defn_name} in derivatives.yaml defines the "
+                                   "forward definition of gradient as element_wise but this only "
+                                   "works for functions with a single differentiable input and a "
+                                   "single differentiable output.")
+            if not len(derivatives) == 1:
+                raise RuntimeError(f"Derivative definition of {defn_name} in derivatives.yaml defines the "
+                                   "forward definition of gradient as element_wise but it does not "
+                                   "defines the gradient formula for its argument which is required.")
+            # This transformation is based on the observation that for element-wise functions, the Jacobian
+            # matrix is diagonal and thus doing J * v or v * J gives the same result.
+            # So here we are going to re-use the backward formula and replace two things:
+            # 1) all occurrences of "grad" with "foo_t", where foo is the name of the unique differentiable input.
+            # 2) all usage of an original input "foo" with its primal value "foo_p".
+            # For example, for abs, the backward formula is:
+            #   grad * self.sgn()
+            # And this function generates a forward formula that is:
+            #   self_t * self_p.sgn()
```

Comment:
People will come along and read the codegen'ed code so it would be nice if the names were a bit more descriptive (e.g. foo_p -> foo_primal, foo_t -> foo_tangent

Reply:
The names here match the ones used in derivatives.yaml.
I don't want to have too long names there to keep the formulas short and readable.